### ISSUE TITLE
The first repeating unpinned Citation Key ends with "b" instead of "a"

### ISSUE THREAD

**Issue by dotran** [retorquere/zotero-better-bibtex#707]
Report ID: 7SSH4AWZ

This is for Zotero 5 + BBT 4050 with **Citation key format** being `[authorsAlpha:lower][year]`.

Accepting the fact that I can only have *unpinned* citekeys for new references, I notice that when a citekey is repeated (under my above short format), the second repeating citekey ends with letter `b` instead of `a`.

For example, the first reference has citekey `ram2015` and the second reference due to its details *should* also have citekey `ram2015`, then we know that the proper citekey for the second reference is `ram2015a`. However, what BBT shows is `ram2015b`.

I hope this will not take much of your time to fix.

**Comment by retorquere** [comment]
:pager: bleep bloop; this is your friendly neighborhood build bot announcing new test build [4052](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.2.4052.issue-707.xpi).

**Comment by retorquere** [comment]
Can you try 4052? It won't fix the problem but it will give me more info about its cause I hope.

**Comment by dotran** [comment]
Yes, I just tried with the 4052. And as you said I don't see any change compared to the previous 4050. Report ID (for the reported items) with 4052 is **TJ74ZDXH**.

Important observation: Even with 4050 (i.e. before I got your message about 4052), when I restarted my Zotero, I saw that the first item `ram2015` became `ram2015a` and the second item became `ram2015`. Weird!

(With "the first item", I mean the item that was added first, i.e. with a smaller `dateAdded`, and the second item is the one that followed. That's why I think it makes sense to have `ram2015` in the first place and then `ram2015a`; not `ram2015` followed by `ram2015b` nor `ram2015a` followed by `ram2015`.)

**Comment by retorquere** [comment]
Oh crap that would mean your citekeys database was regenerated. I hope I can find something on that in the logs, but my schedule is packed for the next few hours.

**Comment by dotran** [comment]
OK, feel free to get back to this when you have a gap. No urgent! ;-).

**Comment by retorquere** [comment]
I do realize your time is as valuable as mine, so I try to capitalize on your availability when I can.

**Comment by retorquere** [comment]
When you start up Zotero, do you see the "Assigning keys" popup every time?

**Comment by dotran** [comment]
I think there is that popup. Below are what I've tried to capture in the order they appeared, but I may have missed something in between.

![image](https://user-images.githubusercontent.com/3706848/29660615-0c7f9ece-88c2-11e7-88ad-275ec1cbfc3b.png)

![image](https://user-images.githubusercontent.com/3706848/29660723-4dacdfec-88c2-11e7-9e09-0e0960212f27.png)

![image](https://user-images.githubusercontent.com/3706848/29660672-2d69201a-88c2-11e7-90bf-f0651033c88b.png)

So the last popup is probably what you expect. This is followed by something like "...Read for Business."

**Comment by retorquere** [comment]
Dammit. Yes, that last one is the popup I meant. You get that at every start?

**Comment by retorquere** [comment]
:pager: bleep bloop; this is your friendly neighborhood build bot announcing new test build [4053](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.2.4053.issue-707.xpi).

**Comment by dotran** [comment]
Yes, I see every time I start Zotero.

**Comment by retorquere** [comment]
Can you try with 4053?

**Comment by retorquere** [comment]
(won't fix, just extra logging)

**Comment by dotran** [comment]
Just tried 4053. Report ID **WFW7IXXI** on the same items as before.

**Comment by retorquere** [comment]
:pager: bleep bloop; this is your friendly neighborhood build bot announcing new test build [4056](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.2.4056.issue-707.xpi).

**Comment by retorquere** [comment]
4056 should take care of the keys being retained at startup. The key skip may be influenced by this -- can you send a new log?

**Comment by dotran** [comment]
Yes, 4056 works as you said: I no more see the "Assigning citation keys" popup from the *second* restart.
New log sent at **DR3NPCDG**

**Comment by retorquere** [comment]
Second restart, right; I could have gone for the first restart but that would have introduced code that would only be run this first instance and never again -- I'd rather not start adding legacy code right now 😄 .

Anyhow -- this build will have added a column named "ID" in the references list which should sort numerically, can you see if you can find item 35107? According to the logs, that reference has claimed ram2015a.

**Comment by dotran** [comment]
No problem with the second restart after installing a new BBT version.

Yes, I can see the "ID" column with item 35107 for ram2015a as in this screenshot

![image](https://user-images.githubusercontent.com/3706848/29678421-cec323fc-88fe-11e7-9306-f3c16573d179.png)

A side note which might be totally unrelated here: When I add a new reference using the "magic stick" **Add Item(s) by Identifier**, I first do not see the Citation Key on the right pane. But *only after* I highlight (i.e. move the focus to) another existing reference and then select back the newly added reference, I see the Citation Key on the right pane. Perhaps this relates to some refreshing mechanism used in the UI... Anyway, this is unimportant, just an observation.

**Comment by retorquere** [comment]
I'm still trying to work out the refresh issue. I have been told I must wait for a specific signal before I can do anything with the db and it's causing all kinds of weirdness. 

Do you still see the postfix skips? If so, with what key?

**Comment by dotran** [comment]
Still have the postfix skip. I've just tried adding a new reference and sent the report **UIMWLFRY**. Screenshots are as follows:

![image](https://user-images.githubusercontent.com/3706848/29685152-dc599440-8914-11e7-98e0-238a5c4edea5.png)

![image](https://user-images.githubusercontent.com/3706848/29685175-effbac68-8914-11e7-8094-0d7862499f7b.png)

**Comment by dotran** [comment]
Do you think this is due to my very citation key format, or should this occur to everybody?

**Comment by retorquere** [comment]
:pager: bleep bloop; this is your friendly neighborhood build bot announcing new test build [4061](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.2.4061.issue-707.xpi).

**Comment by retorquere** [comment]
Your citation key format should not be capable of having this kind of influence, so no, this should happen to everybody -- but it doesn't for me. I can duplicate references to my hearts content and I see them getting consecutive postfixes as expected.

Can you try 4061? I've looked at UIMWLFRY but I don't see any activity indicating new keys are being generated.

**Comment by dotran** [comment]
Just tried 4061. I added two new (duplicated) references. Report ID: **NQ3AWLHN**

Have a look at this screenshot, notice the Date Added and the Citekey -- always postfixes with "b" for the second occurrence of the key.

![image](https://user-images.githubusercontent.com/3706848/29707151-8bf4ee08-8984-11e7-92be-42441f2c800a.png)

**Comment by dotran** [comment]
I am trying adding a third occurrence of the reference, and it has citekey postfixed with "c". So BBT completely omits the "a" postfix.

Also I am trying with Zotero 4 and the corresponding BBT on a Debian VM, works nicely: starting with no postfix, then "a", then "b".

I will make another try with Zotero 5 in Ubuntu VM. Let's see what happens.

**Comment by retorquere** [comment]
I see something in the logs now looks like there's an error when I'm looking whether the key can be kept as-is -- let me look into that.

**Comment by retorquere** [comment]
It was a holdover from when the dynamic key still lived in `extra` -- I have a fix that I think will address the issue lined up in 4063. No idea why I still can't replicate the issue on my end -- must be something to do with the specific order of being called. Can't wait for the proper citekey field to make it to Zotero.

**Comment by dotran** [comment]
OK. By the way, I have the test result for Zotero 5 + BBT on Ubuntu 16.04 VM which is configured just like my Windows 10 but Zotero is completely isolated and fresh, without no references. Now I add the first reference with the magic wand, and repeat that after a few second to have a duplication with a latter timestamp. Same phenomenon:

![image](https://user-images.githubusercontent.com/3706848/29708496-0f7d5ef0-8989-11e7-8476-cd927f34deaf.png)

**Comment by retorquere** [comment]
I believe you -- the behavior is fully explainable given the bug I found. What I can't readily explain is why I was not seeing it.

**Comment by retorquere** [comment]
:pager: bleep bloop; this is your friendly neighborhood build bot announcing new test build [4063](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.2.4063.issue-707.xpi).

**Comment by retorquere** [comment]
And not only am I not seeing it -- a few of the tests I have running generate duplicate keys and should have  failed on that.

**Comment by dotran** [comment]
As you don't have this problem, I just thought there might be something strange existing in my ~13K item database on my working Windows machine. But a fresh install on a VM without any database syncing shows the same sickness.

**Comment by retorquere** [comment]
Even on 4063?!

**Comment by dotran** [comment]
No no, I am trying 4063 now. Just a moment

**Comment by dotran** [comment]
![image](https://user-images.githubusercontent.com/3706848/29708882-a0cc46b8-898a-11e7-8c9d-864b5a657888.png)

So, 4063 does not change the situation. Report at **S7RCVGZ6**

**Comment by retorquere** [comment]
What the *hell*?!

**Comment by dotran** [comment]
The report is generated *after* restarting Zotero. (Then the Citekey column is empty.)

**Comment by retorquere** [comment]
:pager: bleep bloop; this is your friendly neighborhood build bot announcing new test build [4066](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.2.4066.issue-707.xpi).

**Comment by retorquere** [comment]
Oh wait -- you just installed and restarted? That would be as expected -- once assigned, keys don't change unless the reference changes. This is intentional, as you wouldn't want to delete `go2014b` and have `go2014c` change. If you generate new references, it should now generate the postfix sequence `a`, `b`, etc.

The column is empty because Zotero has me waiting forever before I am allowed to access the database. 4066 updates the key list the moment I am allowed to access the database. I'm not fond of this either, but there's nothing I can do about this.

**Comment by dotran** [comment]
Just tried 4066. On Ubuntu VM now works nicely.

In previous tries, I had carefully removed the references (to Trash) and then empty the Trash to completely sweep them out. Then install the new BBT build, then add the references with magic wand.

What I observe now with 4066 is that it works nicely for a new reference, and for references that I delete (to Trash and then delete from Trash also) the issue remains: no postfix --> postfix "b" --> postfix "c".

**Comment by retorquere** [comment]
Wait, I don't understand the second scenario. Can you walk me through this step by step?

**Comment by retorquere** [comment]
I don't see deletion activity in S7RCVGZ6. As a matter of fat I don't see *any* item activity in that log -- I have a notifier set up on all item changes and that spits out a log line -- none of those.

I do see another error in your logs -- apparently I am *also* not allowed to ask where Zotero stores its data until some undocumented promise is resolved. That shouldn't affect this issue you're seeing, but it's another problem to solve before release.

**Comment by dotran** [comment]
OK. Take a look at this

![image](https://user-images.githubusercontent.com/3706848/29709883-72da9206-898e-11e7-8c75-a0197352bd4a.png)

The screenshot shows 4 items, just recently added (as seen by the Date Added column). The `hal2015` and `hall2015a` are brand-new items, and with 4066 we now have correct postfixes. However postfixes are not correct for `go2014` and `go2014b`, which are just added right now (the previous versions of this reference were deleted and they went to Trash, and I have emptied the Trash to completely remove them from the library, that's why the newly added version starts with `go2014`, i.e. no postfix).

**Comment by dotran** [comment]
The log for the above: **ZU6RTWRM**

**Comment by retorquere** [comment]
And once again, no item activity of any kind in your logs. That would explain a whole bunch of issues. I must wait for an answer on [this](https://groups.google.com/forum/#!topic/zotero-dev/klOaKytePA4) because it may very well be fallout from not being allowed to read the data dir in your installation at init -- which would be a timing issue, which would explain why you're seeing it and I'm not.

**Comment by dotran** [comment]
So, I think you've solved the bug for this issue #707.

**Comment by dotran** [comment]
Great!

**Comment by retorquere** [comment]
Yeah, but the other issue is not so great. How long does the popup say "Waiting for Zotero database" approximately before it changes to "BetterBibTeX: Initializing"?

**Comment by dotran** [comment]
I am trying to check what's happening with `go2014` as its first duplication is *always* `go2014b` -- never be `go2014a`, even when I delete all instances of this reference and empty my Trash. This sickness does not exist for other references.

And let's see what I've observed:

![image](https://user-images.githubusercontent.com/3706848/29710876-9764e8a2-8992-11e7-96da-e8d9ca1b9118.png)

![image](https://user-images.githubusercontent.com/3706848/29710902-b080dfc6-8992-11e7-88aa-632713f5d078.png)

Do you notice that as soon as I delete the "Includes index" stuff under the reference, I got the correct citekey postfix? Is this what you would expect? (I would not expect this!)

**Comment by retorquere** [comment]
Certainly not. It would indicate the note is grabbing a cite key.

**Comment by dotran** [comment]
Answer to your question: >15 seconds, probably ~20s. Normal or abnormal?

**Comment by dotran** [comment]
> Certainly not. It would indicate the note is grabbing a cite key.

So another bug? A critical one?

**Comment by retorquere** [comment]
:pager: bleep bloop; this is your friendly neighborhood build bot announcing new test build [4067](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.2.4067.issue-707.xpi).

**Comment by retorquere** [comment]
I don't know -- it's just how long it takes before Zotero tells me it's OK to start doing stuff. I'd say that's pretty long, certainly give that I can't prevent you from interacting with the references before I can start doing my initialization.

And yes, a note grabbing a citekey is a critical problem. I don't know if it does -- it shouldn't, I check for this -- but who knows at this point.

**Comment by retorquere** [comment]
Can you send me a new debug log? and what's the ID of the note? Does it show a citekey?

**Comment by retorquere** [comment]
4067 is an attempt at solving the error in the log ahead of the info I hope to get from [zotero-dev](https://groups.google.com/forum/#!topic/zotero-dev/klOaKytePA4).

I really, really hate the this async crap.

**Comment by retorquere** [comment]
I double-check for notes and attachments even -- the note really should not be grabbing anything, but the debug log and the note id/citekey will tell.

**Comment by dotran** [comment]
![image](https://user-images.githubusercontent.com/3706848/29711456-29a79e74-8995-11e7-9e31-e70090cc0c70.png)

**Comment by dotran** [comment]
Debug log: **3M48948T**

**Comment by dotran** [comment]
What the hell! Look:
![image](https://user-images.githubusercontent.com/3706848/29711619-f75cb1f6-8995-11e7-8ad9-eef2f5293f0f.png)

I have changes anything, but now the `go2014b` becomes `go2014a` magically!

**Comment by retorquere** [comment]
:pager: bleep bloop; this is your friendly neighborhood build bot announcing new test build [4068](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.2.4068.issue-707.xpi).

**Comment by dotran** [comment]
I know why: when a click on the yellow note "Includes index" under `go2014b` and then click black its parent `go2014b` item, the citekey automatically changes to `go2014a`. Maybe you know why it behaves like that.

**Comment by retorquere** [comment]
:pager: bleep bloop; this is your friendly neighborhood build bot announcing new test build [4069](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.2.4069.issue-707.xpi).

**Comment by retorquere** [comment]
At least I now see item activity in your logs. That's a step forwards. The citekeys not updating in the pane, I don't yet know how to solve that, but if it updates after you click away and back, then it's only a refresh issue on the UI (which is a whole other mess).

Can you try 4069? The log will give me timing info on the Zotero startup -- I hope to convince the Zotero devs to plug me in to an earlier phase in the startup. One second is perhaps tolerable, but if it's really a 20 second window where you can do whatever you want with the references before I can even start up, that's really not.

**Comment by retorquere** [comment]
(I meant try and send me a new log)

**Comment by dotran** [comment]
OK, just a moment...

**Comment by dotran** [comment]
> At least I now see item activity in your logs. That's a step forwards. The citekeys not updating in the pane, I don't yet know how to solve that, but if it updates after you click away and back, then it's only a refresh issue on the UI (which is a whole other mess).

I don't know if it's the refresh issue, but if I click on some other items and then click back the `go2014b`, nothing changes. Only when I click on its note and and click back, it changes to the correct `go2014a`.

**Comment by retorquere** [comment]
*sigh* no, that should not happen even knowing the current refresh problems.

**Comment by dotran** [comment]
The 15 second for "Waiting for Zotero database" is also the period that Zotero is loading items, i.e. in its main pane there is no references list shown, only the announcement "Loading items..."

**Comment by retorquere** [comment]
Oh really. Hmm. OK, the log will give me the exact timing.

**Comment by retorquere** [comment]
That click-on-the-note-to-refresh, does that set the key display right for the reference list, or for the details pane?

**Comment by dotran** [comment]
Yes it does.

**Comment by dotran** [comment]
New log: **HQU2SQAC**

**Comment by retorquere** [comment]
I mean which of the two :smile: . Or do they update to the correct key with the exact same behavior? Refresh for the list and for the details are different mechanisms.

**Comment by dotran** [comment]
Ah both: once I click back the parent ref *from* its note, both the Citation Key field on the right pane *and* the value on your Citekey column (next to the ID column) change to the correct "a" postfix.

**Comment by retorquere** [comment]
Hmm, that's not actually *so* bad; a 2.3 second window between the UI being ready and the schema being updated. Perhaps not worth going to battle on.

**Comment by dotran** [comment]
Sure. That's just fine.

**Comment by retorquere** [comment]
Yeah so the refresh is going to be another lovely piece of work, but it is starting to look like #707 may actually be fixed (and thanks a ton for your patience in this).

**Comment by retorquere** [comment]
(still not entirely sure where it got the wrong key from at all)

**Comment by retorquere** [comment]
:pager: bleep bloop; this is your friendly neighborhood build bot announcing new test build [4070](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.2.4070.issue-707.xpi).

**Comment by dotran** [comment]
Now an interesting observation regarding the click-on-the-note: The is no "refresh" if I don't do the "refresh" on the same Zotero *session* that the references are added. In other words, if I add `go2014` (which has a note) and then add it again to make a duplication to get the wrong `go2014b`, and I don't click on the note to rectify the `go2014b`, I just *restart* Zotero, then the click-to-refresh does not work at all -- not like what I'd work if I had done that before restarting Zotero.

**Comment by dotran** [comment]
But that's clearly another issue, and I know how to get the correct postfix for repeated citekeys (of references having a child note): note-clicking on the same session, or forcing key pinning (with `bibtex: ...` on the `Extra` field) after restart.

**Comment by retorquere** [comment]
That would indicate it's actually first assigned `b` and then clicking on the note actually reassigns its parent... wut? I'll go trawling through the log again. Can you send me a fresh log where the click-to-refresh did correct the situation? Sorry, so much things to keep track of, easier for me on a fresh log.

**Comment by retorquere** [comment]
It must be still the same issue. That item should never have been assigned `b`.

**Comment by dotran** [comment]
OK, but what kind of log would you need:
1. Duplicated item just added and I immediately send the log without me performing the note-click-to-correct action.
1. It's added and I have performed a note-click to get the right key and then send the log.
1. It's added and I don't do anything but restart.

**Comment by dotran** [comment]
Also, with build 4070 installed?

**Comment by retorquere** [comment]
Number 2 please. Sure, 4070 is good.

**Comment by dotran** [comment]
OK, just a minute.

**Comment by dotran** [comment]
Please see the log **MQMQQUQ7**

**Comment by dotran** [comment]
BTW, I find it's useful to have the `Citekey` column, so please keep it around. (You might want to hide it by default.)

**Comment by retorquere** [comment]
The citekey column stays, I'll probably move it out of the default columns to be displayed, but the ID column will go.

**Comment by retorquere** [comment]
Which item/note are involved in the click-to-refresh?

**Comment by dotran** [comment]
The `go2014a`. Before click-to-refresh, it was `go2014b`.

**Comment by retorquere** [comment]
:pager: bleep bloop; this is your friendly neighborhood build bot announcing new test build [4072](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.2.4072.issue-707.xpi).

**Comment by retorquere** [comment]
Can you try the same with 4072? I've found the place where it happens but not the reason.

**Comment by dotran** [comment]
Just tried. Log **YVNRP5KF**

**Comment by retorquere** [comment]
OK, that was really, really stupid. Try 4074 when it drops.

Cthulhu I feel dumb now.

**Comment by retorquere** [comment]
:pager: bleep bloop; this is your friendly neighborhood build bot announcing new test build [4074](https://github.com/retorquere/zotero-better-bibtex/releases/download/builds/zotero-better-bibtex-5.0.2.4074.issue-707.xpi).

**Comment by dotran** [comment]
Works! -- the first duplicate citekey postfixes `a`.

**Comment by retorquere** [comment]
My god that was a dumb mistake. Thanks for sticking this one out.

**Comment by dotran** [comment]
Good to know you've sorted this out. Congrats and thanks for the hard work!

**Comment by 7Moses** [comment]
Thanks! for your much-used extension. I have just upgraded to Z5 and extension, Version
5.0.27. The 1+ hour wait of Better BibTex is "Waiting for Zotero database" is not working out for me, ... Any guidance would be appreciated, 

And on this: https://github.com/retorquere/zotero-better-bibtex/blob/d06e541ef7b936e2a6d7b52d8eb7b690aa6403e7/locale/en-US/zotero-better-bibtex.properties,  

Thanks,

Best,

**Comment by retorquere** [comment]
You can submit a BBT debug log from the help menu.

How many references would you guess you have?

**Comment by retorquere** [comment]
Wait, this is not likely related to #707. Can you first upgrade to 5.0.41 and then open a new issue?

**Comment by 7Moses** [comment]
Thanks for the guidance. I enabled logging, restarted Z and Z+BB are playing nice and Snappy, ... Again your script is excellent and saves me a ton of labour, Thanks!

Best,

**Comment by retorquere** [comment]
It's a little more than a script :laughing: 

So is your problem solved now?

**Comment by 7Moses** [comment]
The problem is solved, Thank you. Previously the "waiting for..." was an hour+, now minutes, 
Agree totally ... Your work is praised and appreciated! here, 
Best,
